daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala

```diff
@@ -103,23 +104,35 @@ object Reference {
     override def pretty: String = s"template without contract key $tyCon."
   }
 
-  final case class TemplateImplements(templateName: TypeConName, ifaceName: TypeConName)
+  /** References an interface implementation of interfaceName for templateName,
+    * if a unique one exists.
+    */
+  final case class InterfaceInstance(interfaceName: TypeConName, templateName: TypeConName)
       extends Reference {
-    override def pretty: String = s"template $templateName implementation of interface $ifaceName"
+    override def pretty: String = s"interface instance $interfaceName for $templateName"
   }
 
-  final case class InterfaceCoImplements(templateName: TypeConName, ifaceName: TypeConName)
-      extends Reference {
-    override def pretty: String =
-      s"template $templateName co-implementation of interface $ifaceName"
-  }
-
-  final case class TemplateImplementsOrInterfaceCoImplements(
-      templateName: TypeConName,
-      ifaceName: TypeConName,
+  /** References an interface implementation of interfaceName for templateName
+    * defined in templateName if parentTemplateOrInterface == TorI.Template(())
+    * or in interfaceName if parentTemplateOrInterface == TorI.Interface(())
+    */
+  final case class ConcreteInterfaceInstance(
```

Comment:
this one is only used as a `Context` in `Typing.scala` for better error messages. We don't use it in `lookupX` functions since we never look explicitly for an "interface instance in a template" or an "interface instance in an interface"